client/widmgr/widmgr.go

```diff
@@ -4,80 +4,367 @@
-// allocation.
-func (m *WIDMgr) SignIdentities(minIndex uint64, req []*structs.WorkloadIdentityRequest) ([]*structs.SignedWorkloadIdentity, error) {
-	args := structs.AllocIdentitiesRequest{
-		Identities: req,
-		QueryOptions: structs.QueryOptions{
-			Region:        m.region,
-			MinQueryIndex: minIndex - 1,
-			AllowStale:    true,
-			AuthToken:     m.nodeSecret,
-		},
+// If an error is returned the identities could not be fetched and the renewal
+// goroutine was not started.
+func (m *WIDMgr) Run() error {
+	if len(m.widSpecs) == 0 {
+		m.logger.Debug("no workload identities to retrieve or renew")
+		return nil
 	}
-	reply := structs.AllocIdentitiesResponse{}
-	if err := m.rpc.RPC("Alloc.SignIdentities", &args, &reply); err != nil {
-		return nil, err
+
+	m.logger.Debug("retrieving and renewing workload identities", "num_identities", len(m.widSpecs))
+
+	if err := m.getIdentities(); err != nil {
+		return fmt.Errorf("failed to fetch signed identities: %w", err)
 	}
 
-	if n := len(reply.Rejections); n == 1 {
-		return nil, fmt.Errorf("%d/%d signing request was rejected", n, len(req))
-	} else if n > 1 {
-		return nil, fmt.Errorf("%d/%d signing requests were rejected", n, len(req))
+	go m.renew()
+
+	return nil
+}
+
+// Get retrieves the latest signed identity or returns an error. It must be
+// called after Run and does not block.
+//
+// For retrieving tokens which might be renewed callers should use Watch
+// instead to avoid missing new tokens retrieved by Run between Get and Watch
+// calls.
+func (m *WIDMgr) Get(id cstructs.TaskIdentity) (*structs.SignedWorkloadIdentity, error) {
+	token := m.get(id)
+	if token == nil {
+		// This is an error as every identity should have a token by the time Get
+		// is called.
+		return nil, fmt.Errorf("unable to find token for task %q and identity %q", id.TaskName, id.IdentityName)
 	}
 
-	if len(reply.SignedIdentities) == 0 {
-		return nil, fmt.Errorf("empty signed identity response")
+	return token, nil
+}
+
+func (m *WIDMgr) get(id cstructs.TaskIdentity) *structs.SignedWorkloadIdentity {
+	m.lastTokenLock.Lock()
+	defer m.lastTokenLock.Unlock()
+
+	return m.lastToken[id]
+}
+
+// Watch sends new signed identities until it is closed due to shutdown. Must
+// be called after Run.
+//
+// The caller must call the returned func to stop watching.
+func (m *WIDMgr) Watch(id cstructs.TaskIdentity) (<-chan *structs.SignedWorkloadIdentity, func()) {
+	m.watchersLock.Lock()
+	defer m.watchersLock.Unlock()
+
+	// If Shutdown has been called return a closed chan
+	if m.stopCtx.Err() != nil {
+		c := make(chan *structs.SignedWorkloadIdentity)
+		close(c)
+		return c, func() {}
+	}
+
+	//TODO I think we only want 1 watcher per id, so let's just close on double watches?
+	if existing, ok := m.watchers[id]; ok {
+		close(existing)
+	}
+
+	// Buffer of 1 so sends don't block on receives
+	c := make(chan *structs.SignedWorkloadIdentity, 1)
+	m.watchers[id] = c
```

Comment:
Yeah... I suppose it's up to the watchers to make sure they `cancel()`? Do you have a better solution in mind?

Reply:
> This buffer can only protect the channel from being blocked once right? A watcher that stops listening but doesn't call the `stop` function may leak the goroutine.

I don't think there's any goroutine to leak here? 

In any case, I think it's reasonable to rely on the watcher canceling themselves, especially given that we ensure everything's been closed in the `Shutdown` method. The `stop` function for each watcher deletes the channel from the list (the channel will get GC'd), which prevents the sender (the `renew` goroutine) from blocking on the send because the channel isn't there anymore. We've left it up to `Shutdown` to close and open channels to notify the watcher they can close. So long as the watcher doesn't poll on the channel after it's called `stop`, we should be safe.